### PR TITLE
[ML/Train] Add `TensorflowTrainer` interface

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Interface for `TensorflowTrainer`

Depends on https://github.com/ray-project/ray/pull/22988

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by amogkam — APPROVED** [review]
Nice lgtm!

File: python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py
Comment (richardliaw) [inline python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py:25]: remove typing?

**Review by richardliaw — APPROVED** [review]
lgtm, assuming this is consistent with TorchTrainer

File: python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py
Comment (Yard1) [reply python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py:25]: It's still present in https://github.com/ray-project/ray/pull/22989/

Comment (Yard1) [reply python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py:25]: Will do!

**Comment by ericl** [comment]
Blocked on Trainer base work.

File: python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py
Comment (ericl) [inline python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py:18]: Can we say just the train dataset is sharded?

**Comment by amogkam** [comment]
@Yard1 can we update the docstring to add more details like how to save the model and also an example?

We can use this as reference https://github.com/ray-project/ray/pull/22989

**Comment by Yard1** [comment]
@amogkam done, PTAL!

**Review by krfricke — APPROVED** [review]
Interface looks good

File: python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py
Comment (amogkam) [inline python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py:104]: Need config here since you are using it for config["num_epochs"]!
```suggestion
            def train_loop_for_worker(config):
```

**Review by amogkam — APPROVED** [review]
LGTM! Just one minor comment.

File: python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py
Comment (Yard1) [inline python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py:5]:
```suggestion
from ray.ml.train.data_parallel_trainer import DataParallelTrainer
```

Comment (Yard1) [inline python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py]:
```suggestion
from ray.ml.train.data_parallel_trainer import DataParallelTrainer
```

File: python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py
Comment (Yard1) [inline python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py:130]:
```suggestion
                datasets={"train": train_dataset},
                train_func_config={"num_epochs": 2})
```

File: python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py
Comment (Yard1) [inline python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py:158]:
```suggestion
        train_loop_per_worker: Union[Callable[[], None], Callable[[Dict], None]],
        train_loop_config: Optional[Dict] = None,
```

File: python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py
Comment (Yard1) [inline python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py:1]:
```suggestion
from typing import Callable, Optional, Dict, Union
```

File: python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py
Comment (Yard1) [inline python/ray/ml/train/integrations/tensorflow/tensorflow_trainer.py:142]:
```suggestion
        tensorflow_config: Configuration for setting up the TensorFlow backend. If set to
            None, use the default configuration. This replaces the ``backend_config``
            arg of ``DataParallelTrainer``.
```

**Review by ericl — APPROVED** [review]

**Pull request merged**
